dynamiqs/solvers/utils/utils.py

```diff
@@ -29,12 +29,15 @@ def kraus_map(rho: Tensor, O: Tensor) -> Tensor:
 
     Args:
         rho: Density matrix of shape `(a, ..., n, n)`.
-        operators: Kraus operators of shape `(a, b, n, n)`.
+        operators: Kraus operators of shape `(a, b, c, n, n)` or `(x, a, b, c, n, n)`.
 
     Returns:
         Density matrix of shape `(a, ..., n, n)` with the Kraus map applied.
     """
-    return torch.einsum('abij,a...jk,abkl->a...il', O, rho, O.mH)
+
+    if O.dim() == 5:
+        O = O.unsqueeze(0)
+    return torch.einsum('n...ik,...kl,n...lj->...ij', O, rho, O.mH)
```

Comment:
I don't think this achieves the same operation as described in the docstring. One of the two needs to be fixed.

Reply:
Also, I guess `c` is `b_jump_ops` but I didn't understand what the `x` dimension was?

Reply:
`c` is `b_jump_ops` and `x` is `n_jump_ops`

Reply:
I hotfixed the doc but it's still not perfect. I think we should talk about it IRL

Reply:
Ok, I get it now. So I think we need to have two separate operations:

1. `kraus_map` which computes `torch.einsum('n...ik,...kl,n...ij', O, rho, O.mH)`.
2. `kraus_matmul` which computes `torch.einsum('...ik,...kl,...ij', O, rho, O.mH)`, i.e. without the sum. I think this second one is completely equivalent to `O @ rho @ O.mH`, but to be checked.

In both cases, the `...` clearly designate all batched dimensions which are common to `O` and `rho`. This way, it's more robust if we ever decide to have additional batching dimensions, and we don't have to check `O.dim() == 5`.

Could introduce these two operators and make the appropriate changes? I guess we only need `kraus_matmul` for the product of `M0 @ rho @ M0.mH`.

Reply:
I think it's really better not to use `einsum` when it's not required. It's a complicated function which is particularly inefficient for small-sized systems. At some point we will probably remove all `einsum`s from the code.

So here, using a simple `@` product when it's possible should be favored I think. It also avoids having an `.unsqueeze()` under the hood. This way `kraus_matmul` stays quite straightforward.

Reply:
This goes againt the comment left by @pierreguilmin  in the Kraus map documentation:
```
This is equivalent to `torch.sum(operators @ rho[None,...] @ operators.mH, dim=0)`.
    The use of einsum yields better performances on large matrices, but may cause a
    small overhead on smaller matrices (N <~ 50).
```
I am not litterate on this so until we have a significant benchmark to prove one side or the other I'll leave it as is

Reply:
But here this is not a `torch.sum` that is required, but really just a `@` which is always faster than einsum (at least for square matrices). 

Anyway, I'll fix this in a subsequent PR.